tests/consensus_tests/test_failed_snapshot_recovery.py

```diff
@@ -212,4 +212,84 @@ def test_failed_snapshot_recovery(tmp_path: pathlib.Path):
     new_dense_search_result = search(peer_api_uris[-1], dense_query_vector, query_city)
     assert len(new_dense_search_result) == len(initial_dense_search_result)
     for i in range(len(new_dense_search_result)):
-        assert new_dense_search_result[i]["id"] == initial_dense_search_result[i]["id"]
\ No newline at end of file
+        assert new_dense_search_result[i]["id"] == initial_dense_search_result[i]["id"]
+
+
+@pytest.mark.parametrize("transfer_method", ["snapshot", "stream_records", "wal_delta"])
+def test_dirty_shard_handling_with_active_replicas(tmp_path: pathlib.Path, transfer_method: str):
+    assert_project_root()
+
+    peer_api_uris, peer_dirs, bootstrap_uri = start_cluster(
+        tmp_path,
+        N_PEERS,
+        extra_env={"QDRANT__STORAGE__SHARD_TRANSFER_METHOD": transfer_method },
+    )
+
+    create_collection(
+        peer_api_uris[0], shard_number=N_SHARDS, replication_factor=N_REPLICAS
+    )
+    wait_collection_exists_and_active_on_all_peers(
+        collection_name=COLLECTION_NAME, peer_api_uris=peer_api_uris
+    )
+
+    wait_for_same_commit(peer_api_uris=peer_api_uris)
+
+    upsert_random_points(peer_api_uris[0], 1_000)
+
+    query_city = "London"
+
+    dense_query_vector = random_dense_vector()
+    initial_dense_search_result = search(
+        peer_api_uris[0], dense_query_vector, query_city
+    )
+    assert len(initial_dense_search_result) > 0
+
+    # Create a shard initializing flag to trigger dirty shard fixing logic
+    # this can happen in practice when a node is killed while shard directory was being moved from /qdrant/snapshots to /qdrant/storage
+    # the initializing flag is created but never deleted in such cases, when Qdrant restarts it considers it as dirty shard and tries to recover it
+    flag_path = shard_initializing_flag(peer_dirs[-1], COLLECTION_NAME, 0)
+    Path(flag_path).touch()
+
+    # Kill last peer
+    p = processes.pop()
+    p.kill()
+
+    # Restart same peer
+    peer_api_uris[-1] = start_peer(
+        peer_dirs[-1], f"peer_{N_PEERS}_restarted.log", bootstrap_uri
+    )
```

Comment:
IMO we should also try corrupting the files on disk, and totally removing the local shard data. We want to make sure it still properly starts in those scenarios.

Reply:
Doesn't really matter when you have a dummy shard, but added.